### DISCUSSION TITLE
How can I access target name in a Parcel plugin?

### DISCUSSION THREAD

**Discussion by melted-away** [parcel-bundler/parcel#8252] [Q&A]
I'm using Parcel for a web extension project. I've defined 2 targets in `package.json` file:
```
"targets": {
    "firefox": {
        "engines": {
            "browsers": [
                "last 1 Firefox version"
            ]
        }
    },
    "chrome": {
        "engines": {
            "browsers": [
                "last 1 Chrome version"
            ]
        }
    }
}
```
I want to write a transformer (or any plugin which does the trick) to combine a `manifest.json` and `manifest.{targetName}.json` before collecting manifest dependencies.

So far I have learned that `asset.env` gives some info about the current environment. But I'm looking for a way to get the target name. Is there any way for it?

I've also ask the question on Stackoverflow:
https://stackoverflow.com/questions/72722727/how-can-i-access-target-name-in-a-parcel-plugin

**Comment by melted-away** [comment]
I've come up with a solution:
https://stackoverflow.com/a/72834908/10132738

**Reply by mischnic** [reply]
The reason why the target isn't available in a transformer is that multiple dependencies with different targets could resolve to a single asset. Then there no single target for that asset.

If you're seeing that modifying the package.json doesn't rerun the resolver, then you should list the package.json's file path in the `invalidateOnFileChange` array that can be returned in the resolver.

**Reply by melted-away** [reply]
I did so but didn't get what I expected. So I tried calling the *default resolver*'s `resolve()` inside my _custom resolver_ and then modified the result of the function:
```js
const defaultResolver = require('@parcel/resolver-default').default[Symbol.for('parcel-plugin-config')];
.
.
let res = await defaultResolver.resolve({ specifier, dependency, options });
.
.
res.code = JSON.stringify(manifest);
res.invalidateOnFileChange.push(targetSpecificManifestFilePath);
res.invalidateOnFileCreate.push({ fileName: 'package.json', aboveFilePath: targetSpecificManifestFilePath });
```

But yet changes in `manifest.json` and `manifest.*.json`, while running `npm run watch`, doesn't trigger rebuild. However, all other changes in other files does trigger the rebuild.

**Reply by mischnic** [reply]
> `const defaultResolver = require('@parcel/resolver-default').default[Symbol.for('parcel-plugin-config')];`

The correct way here to is just copy the default resolver's code and modify it. The whole reason everything's abstracted away into `@parcel/node-resolver-core` is to make modifying the resolver easy: https://github.com/parcel-bundler/parcel/blob/v2/packages/resolvers/default/src/DefaultResolver.js

Not sure Why the watching doesn't work. Sharing a more complete resolver would be helpful.

**Reply by melted-away** [reply]
This time I copied the _default resolver_ code into my _custom resolver_. But still no luck.

**Reply by melted-away** [reply]
I've added you to my repository. So that would be great if you could have a look. Thanks!
